ddtrace/contrib/rq/__init__.py

```diff
@@ -0,0 +1,236 @@
+        if instance.is_async and config.rq['distributed_tracing_enabled']:
+            propagator.inject(span.context, job.meta)
+
+        try:
+            return func(*args, **kwargs)
+        finally:
+            # If the queue is not async then the job is run immediately so we can
+            # report the results
+            if not instance.is_async:
+                span.set_tag('job.status', job.get_status())
+
+
+@with_instance_pin
+def traced_queue_fetch_job(rq, pin, func, instance, args, kwargs):
+    with pin.tracer.trace('rq.queue.fetch_job', service=pin.service) as span:
+        span.set_tag('job.id', args[0])
+        return func(*args, **kwargs)
+
+
+@with_instance_pin
+def traced_perform_job(rq, pin, func, instance, args, kwargs):
+    """Trace rq.Worker.perform_job
+    """
+    # `perform_job` is executed in a freshly forked, short-lived instance
+    job = args[0]
+
+    ctx = propagator.extract(job.meta)
+    if ctx.trace_id:
+        pin.tracer.context_provider.activate(ctx)
+
+    try:
+        with pin.tracer.trace('rq.worker.perform_job', service=pin.service, resource=job.func_name) as span:
+            span.set_tag('job.id', job.get_id())
+            return func(*args, **kwargs)
+    finally:
+        if job.get_status() == rq.job.JobStatus.FAILED:
+            span.error = 1
+            span.set_tag('exc_info', job.exc_info)
+        span.set_tag('status', job.get_status())
+        span.set_tag('origin', job.origin)
+
+        # DEV: force flush to agent since the process `os.exit()`s
+        #      immediately after this method returns
+        pin.tracer.writer.stop()
+        pin.tracer.writer.flush_queue()
+
+
+@with_instance_pin
+def traced_job_perform(rq, pin, func, instance, args, kwargs):
+    """Trace rq.Job.perform(...)
+    """
+    job = instance
+
+    with pin.tracer.trace('rq.job.perform', service=pin.service, resource=job.func_name) as span:
+        span.set_tag('job.id', job.get_id())
+        return func(*args, **kwargs)
+
+
+@with_instance_pin
+def traced_job_fetch(rq, pin, func, instance, args, kwargs):
+    """Trace rq.Job.fetch(...)
+    """
+
+    job = None
+    try:
+        with pin.tracer.trace('rq.job.fetch', service=pin.service) as span:
+            span.set_tag('job.id', args[0])
+            job = func(*args, **kwargs)
+            return job
+    finally:
+        if job:
+            job_status = job.get_status()
+            span.set_tag('job.status', job_status)
+
+
+@with_instance_pin
+def traced_job_fetch_many(rq, pin, func, instance, args, kwargs):
+    """Trace rq.Job.fetch_many(...)
+    """
+    with pin.tracer.trace('rq.job.fetch_many', service=pin.service) as span:
+        span.set_tag('job_ids', args[0])
+        return func(*args, **kwargs)
+
+
+def patch():
+    # Avoid importing rq at the module level, eventually will be an import hook
+    import rq
+
+    if getattr(rq, '_datadog_patch', False):
```

Comment:
This is a common pattern we have across all integrations. I wonder if there's a way we can standardize it or if it's just fine to leave it be as is. We could do a context manager or a decorator on `patch`.

Also, are we actually happy with how we do this check? It isn't thread-safe.